java/ql/lib/semmle/code/java/security/InsufficientKeySize.qll

```diff
@@ -0,0 +1,191 @@
+    }
+  }
+
+  /**
+   * A call to the `initialize` method declared in `java.security.KeyPairGenerator`
+   * or to the `init` method declared in `java.security.AlgorithmParameterGenerator`.
+   */
+  private class KeyPairGenInit extends MethodAccess {
+    KeyPairGenInit() {
+      this.getMethod() instanceof KeyPairGeneratorInitMethod or
+      this.getMethod() instanceof AlgoParamGeneratorInitMethod
+    }
+
+    /** Gets the `keysize` argument of this call. */
+    Argument getKeySizeArg() { result = this.getArgument(0) }
+  }
+
+  /**
+   * An instance of a `java.security.KeyPairGenerator`
+   * or of a `java.security.AlgorithmParameterGenerator`.
+   */
+  private class KeyPairGen extends GeneratorAlgoSpec {
+    KeyPairGen() {
+      this instanceof JavaSecurityKeyPairGenerator or
+      this instanceof JavaSecurityAlgoParamGenerator
+    }
+
+    override Expr getAlgoSpec() {
+      result =
+        [
+          this.(JavaSecurityKeyPairGenerator).getAlgoSpec(),
+          this.(JavaSecurityAlgoParamGenerator).getAlgoSpec()
+        ]
+    }
+  }
+}
+
+/** Provides models for symmetric cryptography. */
+private module Symmetric {
+  /** A source for an insufficient key size used in AES algorithms. */
+  private class Source extends InsufficientKeySizeSource {
+    Source() { this.asExpr().(IntegerLiteral).getIntValue() < getMinKeySize() }
+
+    override predicate hasState(DataFlow::FlowState state) { state = getMinKeySize().toString() }
+  }
+
+  /** A sink for an insufficient key size used in AES algorithms. */
+  private class Sink extends InsufficientKeySizeSink {
+    Sink() {
+      exists(KeyGenInit kgInit, KeyGen kg |
+        kg.getAlgoName() = "AES" and
+        DataFlow::localExprFlow(kg, kgInit.getQualifier()) and
+        this.asExpr() = kgInit.getKeySizeArg()
+      )
+    }
+
+    override predicate hasState(DataFlow::FlowState state) { state = getMinKeySize().toString() }
+  }
+
+  /** Returns the minimum recommended key size for AES algorithms. */
+  private int getMinKeySize() { result = 128 }
+
+  /** A call to the `init` method declared in `javax.crypto.KeyGenerator`. */
+  private class KeyGenInit extends MethodAccess {
+    KeyGenInit() { this.getMethod() instanceof KeyGeneratorInitMethod }
+
+    /** Gets the `keysize` argument of this call. */
+    Argument getKeySizeArg() { result = this.getArgument(0) }
+  }
+
+  /** An instance of a `javax.crypto.KeyGenerator`. */
+  private class KeyGen extends GeneratorAlgoSpec instanceof JavaxCryptoKeyGenerator {
+    override Expr getAlgoSpec() { result = JavaxCryptoKeyGenerator.super.getAlgoSpec() }
+  }
+}
+
+/** An instance of a generator that specifies an encryption algorithm. */
+abstract private class GeneratorAlgoSpec extends CryptoAlgoSpec {
+  /** Returns an uppercase string representing the algorithm name specified by this generator object. */
+  string getAlgoName() { result = this.getAlgoSpec().(StringLiteral).getValue().toUpperCase() }
```

Comment:
Would it make sense to use `CompileTimeConstantExpr` here instead?

```suggestion
  string getAlgoName() { result = this.getAlgoSpec().(CompileTimeConstantExpr).getStringValue().toUpperCase() }
```

Reply:
Updated in https://github.com/github/codeql/pull/10785/commits/f40eefce57fda15bdf03953a7d6fcda89070b040. I'm rerunning MRVA to see if this affects the results.